internal/graph/graph.go

```diff
@@ -0,0 +1,209 @@
+	// object reference with a target if the type of the relationship ingress is that
+	// of a TupleToUserset
+	TuplesetRelation *openfgapb.RelationReference
+}
+
+// ConnectedObjectGraph represents a graph of relationships and the connectivity between
+// object and relation references within the graph through direct or indirect relationships.
+// The ConnectedObjectGraph should be used to introspect what kind of relationships between
+// object types can exist.
+type ConnectedObjectGraph struct {
+	typesystem *typesystem.TypeSystem
+}
+
+// BuildConnectedObjectGraph builds an object graph representing the graph of relationships between connected
+// object types either through direct or indirect relationships.
+func BuildConnectedObjectGraph(typesystem *typesystem.TypeSystem) *ConnectedObjectGraph {
+	return &ConnectedObjectGraph{
+		typesystem: typesystem,
+	}
+}
+
+// RelationshipIngresses computes the incoming edges (ingresses) that are possible between the source object and
+// relation and the target user (user or userset).
+//
+// To look up Ingresses(`document#viewer`, `source`), where `source` is an object type with no relation, find the rewrites and types of viewer in the document type:
+// 1. If `source` is a relational type then add `source, direct` to the result.
+// 2. If `objectType#relation` is a type and `objectType#relation` can be a `source` then add `objectType#relation, direct` to the result.
+// 3. If computed userset, say `define viewer as writer`, then recurse on `document#writer, source`.
+// 4. If tuple-to-userset, say, viewer from parent. Go to parent and find its types. In this case, `folder`. Go to `folder` and see if it has a `viewer` relation. If so, recurse on `folder#viewer, source`.
+//
+// To look up Ingresses(`document#viewer`, `folder#viewer`), find the rewrites and relations of viewer in the document type:
+// 1. If `folder#viewer` is a relational type then add `folder#viewer, direct` to the result.
+// 2. If computed userset, say `define viewer as writer`, then recurse on `document#writer, folder#viewer`.
+// 3. If tuple-to-userset, say, viewer from parent. Go to parent and find its related types.
+//  1. If parent's types includes `folder` type, and `folder` contains `viewer` relation then this is exactly a ttu rewrite.
+//  2. Otherwise, suppose the types contains `objectType` which has a relation `viewer`, then recurse on `objectType#viewer, folder#viewer`
+func (g *ConnectedObjectGraph) RelationshipIngresses(target *openfgapb.RelationReference, source *openfgapb.RelationReference) ([]*RelationshipIngress, error) {
+	return g.relationshipIngressesHelper(target, source, map[string]struct{}{})
+}
+
+func (g *ConnectedObjectGraph) relationshipIngressesHelper(target *openfgapb.RelationReference, source *openfgapb.RelationReference, visited map[string]struct{}) ([]*RelationshipIngress, error) {
+	key := createKey(target.GetType(), target.GetRelation())
```

Comment:
No need to call a function just to call a function. If the function were doing more I'd say it may be warranted.

Let's just do `fmt.Sprintf("%s#%s", target.GetType(), target.GetRelation())` here instead.

Reply:
+1

Reply:
Good point. Done. Thanks!